nomad/state/state_store_variables.go

```diff
@@ -511,3 +518,83 @@ func (s *StateStore) VariablesQuotaByNamespace(ws memdb.WatchSet, namespace stri
 	quotaUsed := raw.(*structs.VariablesQuota)
 	return quotaUsed, nil
 }
+
+// VarLockAcquire is the method used to append a lock to a variable, if the
+// variable doesn't exists, it is created.
+func (s *StateStore) VarLockAcquire(idx uint64,
+	req *structs.VarApplyStateRequest) *structs.VarApplyStateResponse {
+	tx := s.db.WriteTxn(idx)
+	defer tx.Abort()
+
+	// Try to fetch the variable.
+	raw, err := tx.First(TableVariables, indexID, req.Var.Namespace, req.Var.Path)
+	if err != nil {
+		return req.ErrorResponse(idx, fmt.Errorf("variable lookup failed: %v", err))
+	}
+
+	// If the variable exist, we must make sure it doesn't hold a lock already
+	if raw != nil {
+		sv := raw.(*structs.VariableEncrypted)
+		if sv.VariableMetadata.Lock != nil {
+			return req.ErrorResponse(idx, errVarAlreadyLocked)
+		}
+
+		// Update the data to avoid overwriting the variable's items while creating the lock
+		req.Var.VariableData = sv.VariableData
+	}
+
+	resp := s.varSetTxn(tx, idx, req)
+	if resp.IsError() {
+		return resp
+	}
+
+	if err := tx.Commit(); err != nil {
+		return req.ErrorResponse(idx, err)
+	}
+
+	return resp
+}
+
+func (s *StateStore) VarLockRelease(idx uint64,
+	req *structs.VarApplyStateRequest) *structs.VarApplyStateResponse {
+	tx := s.db.WriteTxn(idx)
+	defer tx.Abort()
+
+	// Look up the entry in the state store.
+	raw, err := tx.First(TableVariables, indexID, req.Var.Namespace, req.Var.Path)
+	if err != nil {
+		return req.ErrorResponse(idx, fmt.Errorf("failed variable lookup: %s", err))
+	}
+
+	if raw == nil {
+		// Should this be a conflict?
```

Comment:
We could make this and the "lock not found" error below return an error saying it's a conflict, but I think that makes concurrency bugs harder to debug in the caller. Ex. if they delete a variable and then try to release it in a different goroutine, it'll be easier for them to figure out what they did wrong if we say "not found" than calling it a conflict.